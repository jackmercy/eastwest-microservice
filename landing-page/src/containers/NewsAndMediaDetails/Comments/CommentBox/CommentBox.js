import { memo, useCallback, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { useIntl } from 'react-intl';
import { useDispatch } from 'react-redux';
import GoogleLogin from 'react-google-login';
import TextareaAutosize from 'react-textarea-autosize';
// eslint-disable-next-line max-len
import FacebookLogin from 'react-facebook-login/dist/facebook-login-render-props';

import { useOnClickOutside } from '../../../../hooks';
import { Button, PopupMenu } from '../../../../components';
import { FACEBOOK_APP_ID, GOOGLE_CLIENT_ID } from '../../../../../configs';
import { callLoginAsCommunityMember } from '../../../../redux/ducks/news.duck';
import { notifyError } from '../../../../redux/ducks/notification.duck';

import useStyles from './commentBox.styles';

const CommentBox = ({ user, className, onCancel, onSubmit, ...props }) => {
  const buttonRef = useRef();
  const intl = useIntl();
  const classes = useStyles();
  const dispatch = useDispatch();
  const [value, setValue] = useState('');
  const [showMenu, setShowMenu] = useState(false);

  const handleCancelButtonClick = useCallback(() => {
    setValue('');
    onCancel();
  }, [onCancel]);

  const handleInputChange = useCallback(({ target: { value } }) => {
    setValue(value);
  }, []);

  const handleSubmit = useCallback(() => {
    if (!user) {
      setShowMenu(!showMenu);
    } else {
      onSubmit(value, (success) => {
        if (success) {
          setValue('');
        }
      });
    }
  }, [showMenu, user, value]);

  const seperateName = useCallback((name) => {
    const words = name.split(' ');

    return { firstName: words[0], lastName: words[1] };
  }, []);

  const handleSendWithFB = useCallback(
    (response) => {
      if (response && response.email && response.userID) {
        dispatch(
          callLoginAsCommunityMember({
            ...seperateName(response.name),
            avatar: response?.picture?.data?.url,
            emailAddress: response.email,
            socialId: response.userID,
          }),
        );
      } else {
        dispatch(notifyError({ message: 'NEWS.LOGIN.SOCIAL.FAILED' }));
      }
    },
    [dispatch],
  );

  const handleSendWithGG = useCallback(
    (response) => {
      if (response && response.profileObj) {
        dispatch(
          callLoginAsCommunityMember({
            ...seperateName(response.profileObj.name),
            avatar: response.profileObj.imageUrl,
            emailAddress: response.profileObj.email,
            socialId: response.profileObj.googleId,
          }),
        );
      } else {
        dispatch(notifyError({ message: 'NEWS.LOGIN.SOCIAL.FAILED' }));
      }
    },
    [dispatch],
  );

  useOnClickOutside(buttonRef, () => {
    setShowMenu(false);
  });

  return (
    <div className={classNames(classes.root, className)} {...props}>
      <TextareaAutosize
        value={value}
        minRows={3}
        maxRows={20}
        onChange={handleInputChange}
        className={classNames('custom-scrollbar', classes.input)}
        placeholder={intl.formatMessage({
          id: 'NEWS.DETAILS.COMMENT.PLACEHOLDER',
        })}
      />
      <div className={classes.buttonGroup}>
        <Button
          size="small"
          color="tertiary"
          className={classes.button}
          onClick={handleCancelButtonClick}
          dataTest="comment-box-cancel-button"
        >
          {intl.formatMessage({ id: 'NEWS.DETAILS.COMMENT.BUTTON.CANCEL' })}
        </Button>
        <Button
          size="small"
          ref={buttonRef}
          color="primary"
          disabled={!value}
          className={classNames(classes.button, showMenu && 'show')}
          onClick={handleSubmit}
          dataTest="comment-box-submit-button"
        >
          {intl.formatMessage({ id: 'NEWS.DETAILS.COMMENT.BUTTON.SUBMIT' })}
          {!user ? (
            <PopupMenu
              show={showMenu}
              menuPosition="right"
              className={classes.popupMenu}
              list={[
                {
                  key: 'send-with-fb',
                  label: (
                    <FacebookLogin
                      appId={FACEBOOK_APP_ID}
                      fields="name,email,picture"
                      cssClass={classes.testButton}
                      callback={handleSendWithFB}
                      render={(renderProps) => (
                        <button
                          className={classes.socialButton}
                          onClick={renderProps.onClick}
                        >
                          {intl.formatMessage({
                            id: 'NEWS.DETAILS.COMMENT.BUTTON.SUBMIT.FB',
                          })}
                        </button>
                      )}
                    />
                  ),
                },
                {
                  key: 'send-with-gg',
                  label: (
                    <GoogleLogin
                      clientId={GOOGLE_CLIENT_ID}
                      onSuccess={handleSendWithGG}
                      onFailure={handleSendWithGG}
                      cookiePolicy="single_host_origin"
                      isSignedIn={false} // keep user logged in
                      className={classes.testButton}
                      scope="https://www.googleapis.com/auth/userinfo.email"
                      render={(renderProps) => (
                        <button
                          className={classes.socialButton}
                          onClick={renderProps.onClick}
                        >
                          {intl.formatMessage({
                            id: 'NEWS.DETAILS.COMMENT.BUTTON.SUBMIT.GG',
                          })}
                        </button>
                      )}
                    />
                  ),
                },
              ]}
            />
          ) : null}
        </Button>
      </div>
    </div>
  );
};
CommentBox.propTypes = {
  user: PropTypes.shape({
    id: PropTypes.string.isRequired,
    firstName: PropTypes.string.isRequired,
    lastName: PropTypes.string.isRequired,
    avatar: PropTypes.string.isRequired,
  }),
  className: PropTypes.string,
  onCancel: PropTypes.func,
  onSubmit: PropTypes.func,
};
CommentBox.defaultProps = {
  user: undefined,
  className: '',
  onCancel: () => {},
  onSubmit: () => {},
};

export default memo(CommentBox);
