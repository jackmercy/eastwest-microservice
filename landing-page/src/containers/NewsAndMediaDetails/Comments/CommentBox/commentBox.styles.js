import { createUseStyles } from 'react-jss';

import COLORS from '../../../../constants/COLORS';
import { hexToRgbA, rem } from '../../../../utils/jss';
import BREAKPOINTS from '../../../../constants/BREAKPOINTS';

export default createUseStyles({
  root: { marginBottom: rem(32) },
  input: {
    resize: 'none',
    fontFamily: 'inherit',
    width: '100%',
    lineHeight: 'normal',
    fontWeight: 400,
    boxSizing: 'border-box',
    padding: [rem(12), rem(16)],
    color: COLORS.textBlack01,
    backgroundColor: COLORS.bgWhite,
    border: `1px solid ${hexToRgbA(COLORS.borderGray02, 0.5)}`,
    borderRadius: rem(5),
    '&:focus': { outline: 'none', borderColor: COLORS.primary },
    '&::placeholder': { color: COLORS.textGray },
    '-moz-appearance': 'textfield',
    '&::-webkit-outer-spin-button, &::-webkit-inner-spin-button': {
      '-webkit-appearance': 'none',
    },
    '&::-ms-clear, &::-ms-reveal': { display: 'none' },
    '&:-ms-input-placeholder': { color: COLORS.textGray },
    '&:disabled': {
      cursor: 'not-allowed',
      color: COLORS.textGray,
      backgroundColor: COLORS.bgGray02,
    },
  },
  buttonGroup: { textAlign: 'right' },
  button: {
    position: 'relative',
    display: 'inline-block',
    marginTop: rem(28),
    overflow: 'visible',
    '& + $button': { marginLeft: rem(10) },
    '&.primary:not(.disabled).show:hover': { opacity: 1 },
  },
  socialButton: {
    cursor: 'pointer',
    display: 'block',
    width: '100%',
    height: '100%',
    padding: 0,
    border: 'none',
    backgroundColor: 'transparent',
    fontSize: 'inherit',
    lineHeight: 'inherit',
    color: 'inherit',
    textAlign: 'inherit',
  },

  [`@media only screen and (max-width: ${BREAKPOINTS.maxMd}px)`]: {
    root: {},
    input: { fontSize: rem(14) },
    button: { display: 'inline-block', marginTop: rem(16) },
  },
});
