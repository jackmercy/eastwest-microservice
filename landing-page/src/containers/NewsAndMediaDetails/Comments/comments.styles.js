import { createUseStyles } from 'react-jss';

import { rem } from '../../../utils/jss';
import COLORS from '../../../constants/COLORS';
import BREAKPOINTS from '../../../constants/BREAKPOINTS';

export default createUseStyles({
  root: {},
  titleGroup: {
    display: 'flex',
    alignItems: 'center',
    borderTop: `1px solid ${COLORS.borderGray01}`,
    paddingTop: rem(32),
    paddingBottom: rem(32),
  },
  title: {
    fontWeight: 900,
    fontSize: rem(24),
    lineHeight: rem(32),
    margin: 0,
  },
  filterButton: {
    marginLeft: rem(32),
    minWidth: rem(124),
    position: 'relative',
    overflow: 'visible',
    '&.primary:not(.disabled):hover': { opacity: 1 },
    '&:not(.show)': {
      boxShadow: 'none',
      color: COLORS.textBlack01,
      backgroundColor: 'transparent',
    },
    '& > [class^=icon-]': { fontSize: rem(16), marginRight: rem(12) },
  },

  [`@media only screen and (max-width: ${BREAKPOINTS.maxSm}px)`]: {
    root: {},
    titleGroup: {
      justifyContent: 'space-between',
      border: 'none',
      paddingTop: 0,
      paddingBottom: rem(24),
    },
    title: { fontSize: rem(18) },
  },
});
