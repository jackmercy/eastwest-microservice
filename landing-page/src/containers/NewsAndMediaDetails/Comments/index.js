import dynamic from 'next/dynamic';

export default dynamic(() => import('./Comments'), { ssr: false });
