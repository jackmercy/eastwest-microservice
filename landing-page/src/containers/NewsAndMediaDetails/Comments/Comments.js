import { memo, useCallback, useMemo, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { useIntl } from 'react-intl';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import { useMediaQuery } from 'react-responsive';

import { Button, PopupMenu } from '../../../components';
import { useOnClickOutside } from '../../../hooks';
import BREAKPOINTS from '../../../constants/BREAKPOINTS';
import {
  getCommunityMember,
  callCreateComment,
} from '../../../redux/ducks/news.duck';
import { storageUtil } from '../../../utils';

import CommentBox from './CommentBox';
import CommentPreview from './CommentPreview';
import EmptyComment from './EmptyComment';
import useStyles from './comments.styles';

const Comments = ({ newsId, data, className }) => {
  const filterButtonRef = useRef();
  const intl = useIntl();
  const classes = useStyles();
  const dispatch = useDispatch();
  const [comments, setComments] = useState(data);
  const [showMenu, setShowMenu] = useState(false);
  const member = useSelector(getCommunityMember, shallowEqual);
  const auth = useMemo(() => {
    const auth = storageUtil.getAuth();
    if (!auth) return member;
    return auth ? JSON.parse(auth) : undefined;
  }, [member]);
  const isMobile = useMediaQuery({
    query: `(max-width: ${BREAKPOINTS.maxSm}px)`,
  });

  useOnClickOutside(filterButtonRef, () => {
    setShowMenu(false);
  });

  const handleFilterButtonClick = useCallback(() => {
    setShowMenu(!showMenu);
  }, [showMenu]);

  const handleFilterItemClick = useCallback(() => {
    setShowMenu(!showMenu);
  }, [showMenu]);

  const handleSubmitComment = useCallback(
    (comment, parent, callback) => {
      dispatch(
        callCreateComment({
          data: {
            user: auth.id,
            news: newsId,
            comment,
            parent,
          },
          callback: (success, data) => {
            if (success && data.status === 'approved') {
              const newComments = [...comments];
              if (data.parent) {
                newComments[data.parent].unshift({ ...data, user: auth });
              } else {
                newComments.unshift({ ...data, user: auth });
              }
              setComments(newComments);
              callback(true);
            }
          },
        }),
      );
    },
    [dispatch, comments],
  );

  return (
    <div className={classNames(classes.root, className)}>
      <div className={classes.titleGroup}>
        <h2 className={classes.title} data-test="comments-title">
          {comments.length}&nbsp;
          {intl.formatMessage({ id: 'NEWS.DETAILS.COMMENT.TITLE' })}
        </h2>
        {comments.length > 0 ? (
          <Button
            ref={filterButtonRef}
            size="small"
            color="primary"
            onClick={handleFilterButtonClick}
            dataTest="comments-filter"
            className={classNames(classes.filterButton, showMenu && 'show')}
          >
            <i className="icon-align-left" />
            {intl.formatMessage({ id: 'NEWS.DETAILS.COMMENT.FILTER' })}
            <PopupMenu
              show={showMenu}
              menuPosition={isMobile ? 'right' : 'left'}
              className={classes.popupMenu}
              list={[
                {
                  key: 'filter-top-comments',
                  label: intl.formatMessage({
                    id: 'NEWS.DETAILS.COMMENT.FILTER.TOP',
                  }),
                  onClick: handleFilterItemClick,
                },
                {
                  key: 'filter-newest-comments',
                  label: intl.formatMessage({
                    id: 'NEWS.DETAILS.COMMENT.FILTER.NEWEST',
                  }),
                  onClick: handleFilterItemClick,
                },
              ]}
            />
          </Button>
        ) : null}
      </div>
      <CommentBox
        id="main-comment-box"
        onSubmit={handleSubmitComment}
        user={auth}
      />
      {comments.length > 0 ? (
        comments.map((comment) => (
          <CommentPreview key={comment.id} data={comment}>
            {comment.replies.length > 0
              ? comment.replies.map((reply) => (
                <CommentPreview
                  key={reply.id}
                  data={reply}
                  onSubmit={(value) => {
                    handleSubmitComment(value, comment.id);
                  }}
                />
              ))
              : null}
          </CommentPreview>
        ))
      ) : (
        <EmptyComment />
      )}
    </div>
  );
};
Comments.propTypes = {
  newsId: PropTypes.string.isRequired,
  data: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      user: PropTypes.shape({
        id: PropTypes.string.isRequired,
        avatar: PropTypes.string.isRequired,
        firstName: PropTypes.string.isRequired,
        lastName: PropTypes.string.isRequired,
      }).isRequired,
      createdAt: PropTypes.string.isRequired,
      updatedAt: PropTypes.string.isRequired,
      comment: PropTypes.string.isRequired,
      likes: PropTypes.array.isRequired,
      replies: PropTypes.array,
    }),
  ),
  className: PropTypes.string,
};
Comments.defaultProps = {
  data: [],
  className: '',
};

export default memo(Comments);
