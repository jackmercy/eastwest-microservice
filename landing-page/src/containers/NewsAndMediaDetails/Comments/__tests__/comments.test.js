import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Comments from '../Comments';
import CommentBox from '../CommentBox';

describe('Testing NewsAndMediaDetails > Comments', () => {
  let wrapper;
  const props = {
    // TODO integration
    data: [
      {
        id: 'abc',
        avatar:
          // eslint-disable-next-line max-len
          'https://play-lh.googleusercontent.com/IeNJWoKYx1waOhfWF6TiuSiWBLfqLb18lmZYXSgsH1fvb8v1IYiZr5aYWe0Gxu-pVZX3',
        firstName: 'Hai',
        lastName: 'Bui',
        updatedAt: '2021-08-03T03:03:13.945Z',
        // eslint-disable-next-line max-len
        comment: `Anim duis cupidatat aliquip id et. Lorem nulla aliqua nisi dolore
    voluptate pariatur deserunt ad est. Sint exercitation minim labore
    aute deserunt cillum laboris aute esse ut.`,
        likes: 100,
        replies: [
          {
            id: 'abcd',
            avatar:
              // eslint-disable-next-line max-len
              'https://play-lh.googleusercontent.com/IeNJWoKYx1waOhfWF6TiuSiWBLfqLb18lmZYXSgsH1fvb8v1IYiZr5aYWe0Gxu-pVZX3',
            firstName: 'Hai',
            lastName: 'Bui',
            updatedAt: '2021-08-03T03:03:13.945Z',
            // eslint-disable-next-line max-len
            comment: `Anim duis cupidatat aliquip id et. Lorem nulla aliqua nisi dolore
    voluptate pariatur deserunt ad est. Sint exercitation minim labore
    aute deserunt cillum laboris aute esse ut.`,
            likes: 100,
            replies: [],
          },
          {
            id: 'abcde',
            avatar:
              // eslint-disable-next-line max-len
              'https://play-lh.googleusercontent.com/IeNJWoKYx1waOhfWF6TiuSiWBLfqLb18lmZYXSgsH1fvb8v1IYiZr5aYWe0Gxu-pVZX3',
            firstName: 'Hai',
            lastName: 'Bui',
            updatedAt: '2021-08-03T03:03:13.945Z',
            // eslint-disable-next-line max-len
            comment: `Anim duis cupidatat aliquip id et. Lorem nulla aliqua nisi dolore
    voluptate pariatur deserunt ad est. Sint exercitation minim labore
    aute deserunt cillum laboris aute esse ut.`,
            likes: 100,
            replies: [],
          },
        ],
      },
    ],
  };

  beforeAll(() => {
    wrapper = mount(<Comments {...props} />);
  });

  it('should render match snapshot', () => {
    const wrapper = shallow(<Comments {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should have a title', () => {
    expect(wrapper.find('[data-test="comments-title"]')).toHaveLength(1);
  });

  it('should have a filter', () => {
    expect(wrapper.find('[data-test="comments-filter"]')).toHaveLength(1);
  });

  it('should have CommentBox', () => {
    expect(wrapper.find(CommentBox)).toHaveLength(1);
  });
});
