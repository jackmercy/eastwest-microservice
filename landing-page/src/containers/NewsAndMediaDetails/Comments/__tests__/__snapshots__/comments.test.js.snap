// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing NewsAndMediaDetails > Comments should render match snapshot 1`] = `
<div
  className="root-0-2-1"
>
  <div
    className="titleGroup-0-2-2"
  >
    <h2
      className="title-0-2-3"
      data-test="comments-title"
    >
      1
      Â 
      Comments
    </h2>
    <Memo([object Object])
      className="filterButton-0-2-4"
      color="primary"
      dataTest="comments-filter"
      onClick={[Function]}
      size="small"
    >
      <i
        className="icon-align-left"
      />
      Filter by
      <Memo([object Object])
        list={
          Array [
            Object {
              "key": "filter-top-comments",
              "label": "Top comments",
              "onClick": [Function],
            },
            Object {
              "key": "filter-newest-comments",
              "label": "Newest comments",
              "onClick": [Function],
            },
          ]
        }
        menuPosition="left"
        show={false}
      />
    </Memo([object Object])>
  </div>
  <Memo(CommentBox)
    id="main-comment-box"
  />
  <Memo(CommentPreview)
    data={
      Object {
        "avatar": "https://play-lh.googleusercontent.com/IeNJWoKYx1waOhfWF6TiuSiWBLfqLb18lmZYXSgsH1fvb8v1IYiZr5aYWe0Gxu-pVZX3",
        "comment": "Anim duis cupidatat aliquip id et. Lorem nulla aliqua nisi dolore
    voluptate pariatur deserunt ad est. Sint exercitation minim labore
    aute deserunt cillum laboris aute esse ut.",
        "firstName": "Hai",
        "id": "abc",
        "lastName": "Bui",
        "likes": 100,
        "replies": Array [
          Object {
            "avatar": "https://play-lh.googleusercontent.com/IeNJWoKYx1waOhfWF6TiuSiWBLfqLb18lmZYXSgsH1fvb8v1IYiZr5aYWe0Gxu-pVZX3",
            "comment": "Anim duis cupidatat aliquip id et. Lorem nulla aliqua nisi dolore
    voluptate pariatur deserunt ad est. Sint exercitation minim labore
    aute deserunt cillum laboris aute esse ut.",
            "firstName": "Hai",
            "id": "abcd",
            "lastName": "Bui",
            "likes": 100,
            "replies": Array [],
            "updatedAt": "2021-08-03T03:03:13.945Z",
          },
          Object {
            "avatar": "https://play-lh.googleusercontent.com/IeNJWoKYx1waOhfWF6TiuSiWBLfqLb18lmZYXSgsH1fvb8v1IYiZr5aYWe0Gxu-pVZX3",
            "comment": "Anim duis cupidatat aliquip id et. Lorem nulla aliqua nisi dolore
    voluptate pariatur deserunt ad est. Sint exercitation minim labore
    aute deserunt cillum laboris aute esse ut.",
            "firstName": "Hai",
            "id": "abcde",
            "lastName": "Bui",
            "likes": 100,
            "replies": Array [],
            "updatedAt": "2021-08-03T03:03:13.945Z",
          },
        ],
        "updatedAt": "2021-08-03T03:03:13.945Z",
      }
    }
    key="abc"
  >
    <Memo(CommentPreview)
      data={
        Object {
          "avatar": "https://play-lh.googleusercontent.com/IeNJWoKYx1waOhfWF6TiuSiWBLfqLb18lmZYXSgsH1fvb8v1IYiZr5aYWe0Gxu-pVZX3",
          "comment": "Anim duis cupidatat aliquip id et. Lorem nulla aliqua nisi dolore
    voluptate pariatur deserunt ad est. Sint exercitation minim labore
    aute deserunt cillum laboris aute esse ut.",
          "firstName": "Hai",
          "id": "abcd",
          "lastName": "Bui",
          "likes": 100,
          "replies": Array [],
          "updatedAt": "2021-08-03T03:03:13.945Z",
        }
      }
      key="abcd"
    />
    <Memo(CommentPreview)
      data={
        Object {
          "avatar": "https://play-lh.googleusercontent.com/IeNJWoKYx1waOhfWF6TiuSiWBLfqLb18lmZYXSgsH1fvb8v1IYiZr5aYWe0Gxu-pVZX3",
          "comment": "Anim duis cupidatat aliquip id et. Lorem nulla aliqua nisi dolore
    voluptate pariatur deserunt ad est. Sint exercitation minim labore
    aute deserunt cillum laboris aute esse ut.",
          "firstName": "Hai",
          "id": "abcde",
          "lastName": "Bui",
          "likes": 100,
          "replies": Array [],
          "updatedAt": "2021-08-03T03:03:13.945Z",
        }
      }
      key="abcde"
    />
  </Memo(CommentPreview)>
</div>
`;
